kqueue.c

```diff
@@ -211,8 +212,9 @@ kq_build_changes_list(const struct event_changelist *changelist,
 			int newsize = kqop->changes_size * 2;
 			struct kevent *newchanges;
 
-			if (newsize < 0 || (size_t)newsize >
-			    EV_SIZE_MAX / sizeof(struct kevent)) {
+			if (kqop->changes_size > INT_MAX / 2 ||
+			    (size_t)newsize > EV_SIZE_MAX /
+			    sizeof(struct kevent)) {
```

Comment:
I think that simply `(kqop->changes_size > INT_MAX / 2)` is enough (before `newsize` assignment). And we do not need `(size_t)newsize > EV_SIZE_MAX` since we already limit size to `INT_MAX`

Reply:
We need the second check for 32 bit systems. Otherwise we could end up with an integer overflow due to multiplication with sizeof(struct kevent) for the mm_realloc argument.

Reply:
>Otherwise we could end up with an integer overflow due to multiplication with sizeof(struct kevent) for the mm_realloc argument.

You right, but please write it against `changes_size` too

*Doh, read that condition incorrectly*

Reply:
That check would have to take changes_size * 2 into account then. Is that more readable?
If a signed integer overflow occurred, the second condition is not checked. If it didn't occur, then the cast is safe and shows that mm_realloc-argument is properly checked.

All in all it shows that this small piece of code is not trivial to review. Maybe a fixed maximum size is an easier solution here? Something like a "#define MAX_EVENTS 8096", you get the idea.

Reply:
> Is that more readable?

The reason I was asking about this is that there is still UB happened, since check was done after it (`newsize` assignment). Yes this bit of code should works correctly now, but someone may skip the part after assignment of `newsize` (overflow check), considering that `newsize` is valid and it can be used in code (it is unlikely in this particular case, but this is about overflow checks in general).

> All in all it shows that this small piece of code is not trivial to review

Kind of, but this is due to two conditions in one places (and because overflow checks is not regular code).

>Maybe a fixed maximum size is an easier solution here? 

I don't know what is the sane limit for kqueue, so I think that shouldn't add such quirks.

Reply:
Adjusted, with the assignment happening after the checks.